### PR TITLE
[IPC4]Dynamic module library loading feature

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/sof/ipc4-loader.c [hunk 1]
```diff
@@ -211,8 +288,81 @@ static void sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
...
+	/* ID 0 is reserved for base FW */
+	id += 1;
+
+	if (id >= ipc4_data->max_fw_libs) {
+		dev_err(sdev->dev, "Library ID: %u exceeds max value %u\n", id,
+			ipc4_data->max_fw_libs);
+		return -EBUSY;
+	}
+
+	ret = request_firmware(&fw, fw_filename, sdev->dev);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:332]: Where are we going to release the firmware?

File: include/sound/sof/ipc4/header.h [hunk 2]
```diff
@@ -361,6 +361,21 @@ enum sof_ipc4_fw_config_params {
...
+	SOF_IPC4_FW_CFG_MAX_LL_TASKS_PER_PRI_COUNT,
+	SOF_IPC4_FW_CFG_LL_PRI_COUNT,
+	SOF_IPC4_FW_CFG_MAX_DP_TASKS_COUNT,
+	SOF_IPC4_FW_CFG_MAX_LIBS_COUNT,
+	SOF_IPC4_FW_CFG_SCHEDULER_CONFIG,
+	SOF_IPC4_FW_CFG_XTAL_FREQ_HZ,
+	SOF_IPC4_FW_CFG_CLOCKS_CONFIG,
+	SOF_IPC4_FW_CFG_RESERVED,
+	SOF_IPC4_FW_CFG_POWER_GATING_POLICY,
+	SOF_IPC4_FW_CFG_ASSERT_MODE,
```
Comment (ujfalusi) [inline include/sound/sof/ipc4/header.h:382]: Should be a fixup! ?

File: sound/soc/sof/ipc4-loader.c [hunk 3]
```diff
@@ -186,6 +187,13 @@ static void sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 		case SOF_IPC4_FW_CFG_TRACE_LOG_BYTES:
 			dev_vdbg(sdev->dev, "Trace log size: %u\n", *tuple->value);
 			break;
+		case SOF_IPC4_FW_CFG_MAX_LIBS_COUNT:
+			ipc4_data->max_fw_libs = *tuple->value;
+			ipc4_data->fw_lib_names = devm_kcalloc(sdev->dev, ipc4_data->max_fw_libs,
+							       LIBRARY_FILENAME_LEN, GFP_KERNEL);
+
+			/* TODO: check if mem alloc is successful and return err */
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:279]: Should we fixup! the fw_loader_ops and make the `query_fw_configuration()` to return `int` to be able to fail?

File: sound/soc/sof/ipc4-loader.c [hunk 4]
```diff
@@ -163,8 +162,10 @@ static void sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 		return;
 
 	ret = iops->set_get_data(sdev, &msg, msg.data_size, false);
-	if (ret)
+	if (ret) {
+		kfree(msg.data_ptr);
 		return;
+	}
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:251]: I have done this w/ goto out, and the out is just before the kfree at the end.

File: sound/soc/sof/ipc4-loader.c [hunk 5]
```diff
@@ -211,8 +288,81 @@ static void sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
...
+	size_t fw_offset;
+	int ret, i;
+	u32 id;
+
+	/* nothing to do if module is part of the base firmware */
+	for (i = 0; ipc4_data->num_fw_modules; i++) {
+		guid_t guid, widget_guid;
+
+		memcpy(&guid, ipc4_data->base_fw_module_uuids[i], sizeof(guid_t));
+		memcpy(&widget_guid, uuid, sizeof(guid_t));
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:306]: What if we would have used guid_t type from the beginning?

File: sound/soc/sof/ipc4-loader.c [hunk 6]
```diff
@@ -211,8 +288,81 @@ static void sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
...
+
+		if (guid_equal(&guid, &widget_guid))
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:308]: Or cast it here only saving on memcpy?

File: sound/soc/sof/sof-priv.h [hunk 7]
```diff
@@ -183,6 +183,7 @@ struct snd_sof_dsp_ops {
 	int (*load_firmware)(struct snd_sof_dev *sof_dev); /* mandatory */
 	int (*load_module)(struct snd_sof_dev *sof_dev,
 			   struct snd_sof_mod_hdr *hdr); /* optional */
+	int (*load_library)(struct snd_sof_dev *sdev, struct firmware *fw, u32 id); /* optional */
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h:186]: what is the difference between `load_module` and `load_library`?

Would we want to unload to make it dynamic?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@ranj063, this is pretty much rewrites the existing fw_loader support which is merged, so most of these (which touches generic code should be fixup).

There are couple of things not too clear for me, can you clarify those.

I will delay the IPC4 fw loader PR in light of this...

File: include/sound/sof/ipc4/header.h (hunk 2 above)
Comment (ranj063) [reply include/sound/sof/ipc4/header.h:382]: yep fixup. will change

File: sound/soc/sof/sof-priv.h (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/sof-priv.h:186]: this is a question for @plbossart. I'm not sure about load_module.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h:186]: I have no idea @ujfalusi @ranj063, I don't think anyone ever used this, we always use the default, don't we?

````c
        load_module = sof_ops(sdev)->load_module;
	if (!load_module) {
		dev_dbg(sdev->dev, "%s: Using generic module loading\n", __func__);
		load_module = sof_ipc3_parse_module_memcpy;
	} else {
		dev_dbg(sdev->dev, "%s: Using custom module loading\n", __func__);
	}
 [...]
	ret = load_module(sdev, module);
````

@lgirdwood any memories of what the intent was on this one?

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:186]: I got the answer from Cezary. Apparently in older platform the library and the module code were separate but newer ones CNL onwards dont have that.

File: sound/soc/sof/ipc4-loader.c (hunk 5 above)
Comment (libinyang) [inline sound/soc/sof/ipc4-loader.c:306]: No need to copy uuid to widget_guid in every iteration.

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -1896,6 +1896,23 @@ static int sof_complete(struct snd_soc_component *scomp)
 			}
 		}
 
+	/* load all 3rd party module libraries */
```
Comment (libinyang) [inline sound/soc/sof/topology.c:1899]: As it is a loadable module, should we load it when necessary and unload it when stream is closed?

File: include/sound/sof/ipc4/header.h (hunk 2 above)
Comment (ujfalusi) [reply include/sound/sof/ipc4/header.h:382]: I think it is not changed in the PR (the patch title)?

File: sound/soc/sof/sof-priv.h (hunk 7 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h:186]: @plbossart, only the default load_module (`sof_ipc3_parse_module_memcpy()`) is used, all architectures dropped setting the sdev level load_module, all is internal to ipc3 now.

File: sound/soc/sof/intel/hda-loader.c [hunk 9]
```diff
@@ -388,6 +389,58 @@ static int hda_dsp_boot_imr(struct snd_sof_dev *sdev)
...
+	ret = sof_ipc_tx_message(sdev->ipc, &msg, 0, NULL, 0);
+
+	ret1 = cl_trigger(sdev, hext_stream, SNDRV_PCM_TRIGGER_STOP);
+	if (ret1 < 0) {
+		dev_err(sdev->dev, "%s: DMA trigger stop failed\n", __func__);
+		if (!ret)
+			ret = ret1;
+	}
+
+	/* clean up even in case of error and return the first error */
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-loader.c:431]: You mean clean up regardless of an error, right?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@ranj063, did you updated this PR? I have strong déjà vu feeling that not much changed.

In my view:
`module_loading()` to load SOF IPC3 modules into the firmware. A 'module' is a set of 'blocks', but it must have at least one. The 'blocks' are loaded to a specific offset to the DSP. The module raw data contains the `struct snd_sof_mod_hdr` and each block contains `struct snd_sof_blk_hdr`.
These modules are within the firmware image itself, but in theory they could be loaded from a separate file, I don't see anything disallowing that.
**Module loading is done before the firmware is booted!**

`load_library()` to load IPC4 library into the firmware and it is only applicable in case of external, modules, they are not in the firmware image. Right?
It is just a binary 'blob' and we set up a DMA to send the module data to the DSP.
**Library loading is done after the firmware is booted!**

The IPC3 module loading is generic among architectures, the abstraction is at `snd_sof_dsp_block_write()` level (so essentially same for all supported arch atm).

The IPC4 library loading is heavily platform dependent, the Intel (and only implementation so far) uses HDA DMA with and an IPC4 message to load a module.

In any case, the module/library loading is platform and IPC specific.

I would change the `load_module`:
- int (*load_module)(struct snd_sof_dev *sof_dev, struct snd_sof_mod_hdr *hdr);
+int (*load_module)(struct snd_sof_dev *sof_dev, void module_data);

and deal with the module_data accoring to the IPC version.
For IPC4 you would need a new container in ipc4-priv.h for a library, it can be simply:
```c
struct sof_ipc4_library {
	u32 id;
	size_t size;
	void *data;
};
```

I would also argue that the `struct sof_ipc_fw_loader_ops` level `load_library`/`get_modues_info` is not needed, this can be handled within ipc4 code just fine?

My point is: what happens if IPC5 will have `addons` or `plugins`? Are they going to be modules or libraries? I think module is a bit more generic to be applied to wide range of loadable code, feature.

**Comment by ranj063** [comment]
need to close for now. Will reopen soon

**Pull request closed without merging**